packages/protobuf/src/transform/index.ts

```diff
@@ -0,0 +1,933 @@
+   */
+  function toMethodFromOperation(operation: Operation): ProtoMethodDeclaration {
+    const streamingMode = program.stateMap(state.stream).get(operation) ?? StreamingMode.None;
+
+    return {
+      kind: "method",
+      stream: streamingMode,
+      name: capitalize(operation.name),
+      input: addImportSourceForProtoIfNeeded(
+        program,
+        addInputParams(operation.parameters, operation),
+        operation,
+        operation.parameters
+      ),
+      returns: addImportSourceForProtoIfNeeded(
+        program,
+        addReturnType(operation.returnType, operation),
+        operation,
+        operation.returnType as NamespaceTraversable
+      ),
+    };
+  }
+
+  /**
+   * Checks a parameter Model satisfies the constraints for a Protobuf method input and adds it to the declarations,
+   * returning a ProtoRef to the generated named message.
+   *
+   * @param model - the model to add
+   * @returns a reference to the model's message
+   */
+  function addInputParams(paramsModel: Model, operation: Operation): ProtoRef {
+    const effectiveModel = computeEffectiveModel(
+      paramsModel,
+      capitalize(operation.name) + "Request"
+    );
+
+    /* c8 ignore start */
+
+    // Not sure if this can or can't actually happen at runtime, but we'll defensively handle it anyway.
+    if (!effectiveModel) {
+      reportDiagnostic(program, {
+        code: "unsupported-input-type",
+        messageId: "unconvertible",
+        target: paramsModel,
+      });
+
+      return unreachable("unsupported input type");
+    }
+    /* c8 ignore stop */
+
+    return checkExtern(effectiveModel, operation);
+  }
+
+  /**
+   * Returns an extern ref if the given type is an instance of `Extern`, otherwise returns a ref to the model's name.
+   */
+  function checkExtern(model: Model, relativeSource: Model | Operation): ProtoRef {
+    const extern = program.stateMap(state.externRef).get(model) as [string, string] | undefined;
+    if (extern) {
+      typeWantsImport(program, relativeSource, extern[0]);
+      return ref(extern[1]);
+    }
+
+    return ref(model.name);
+  }
+
+  /**
+   * Checks that a return type is a Model and converts it to a message, adding it to the declarations and returning
+   * a reference to its name.
+   *
+   * @param t - the model to add
+   * @param operationName - the name of the originating operation, used to compute a synthetic model name if required
+   * @returns a reference to the model's message
+   */
+  function addReturnType(t: Type, operation: Operation): ProtoRef {
+    switch (t.kind) {
+      case "Model":
+        return addReturnModel(t, operation);
+      case "Intrinsic":
+        switch (t.name) {
+          case "unknown":
+            // We just use the built-in "unknown" type to mean the protobuf Any type
+            return ref("Any");
+          case "void": {
+            const emptyRef = program.resolveTypeReference("TypeSpec.Protobuf.Empty")[0];
```

Comment:
should that one not be cached too?

Reply:
I refactored handling of intrinsics a little bit to make it much more robust. We just map `unknown` to `google.protobuf.Any` and `void` to `google.protobuf.Empty` by resolving `WellKnown.Any` and `WellKnown.Empty`, and we cache them at the program level just like we do with scalars now.